base/libgit2/types.jl

```diff
@@ -612,13 +631,60 @@ function getobjecttype(obj_type::Cint)
     end
 end
 
-type UserPasswordCredentials <: AbstractPayload
+"Empty credentials"
+type EmptyCredentials <: AbstractCredentials end
+
+"Credentials that support only `user` and `password` parameters"
+type UserPasswordCredentials <: AbstractCredentials
     user::AbstractString
     pass::AbstractString
-    used::Bool
-    UserPasswordCredentials(u::AbstractString,p::AbstractString) = new(u,p,false)
-end
-user(p::UserPasswordCredentials) = p.user
-password(p::UserPasswordCredentials) = p.pass
-isused(p::UserPasswordCredentials) = p.used
-setused!(p::UserPasswordCredentials, val::Bool) = (p.used = val)
+    count::Int            # authentication failure protection count
+    use_ssh_agent::UInt8  # used for ssh-agent authentication
+    UserPasswordCredentials(u::AbstractString,p::AbstractString) = new(u,p,3,zero(UInt8))
+end
+"Checks if credentials were used or failed authentication, see `LibGit2.credentials_callback`"
+function isused(p::UserPasswordCredentials)
+    p.count <= 0 && return true
+    p.count -= 1
```

Comment:
given this is mutating, it should be either `isused!`, or probably better, `checkused!`